File: php/leet_code/2816.php
<?php
declare(strict_types=1);

/**
 * Copyright ©2012-2024 Jerod Gawne <https://github.com/jerodg/>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License (SSPL) as
 * published by MongoDB, Inc., either version 1 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * SSPL for more details.
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * You should have received a copy of the SSPL along with this program.
 * If not, see <https://www.mongodb.com/licensing/server-side-public-license>.
 */

/**
 * Class ListNode
 *
 * This class represents a node in a singly linked list. Each node contains a value and a reference to the next node.
 */
class ListNode
{
    /**
     * @var mixed The value stored in the node.
     */
    public mixed $val = 0;

    /**
     * @var ListNode|null The next node in the linked list.
     */
    public ?ListNode $next = null;

    /**
     * ListNode constructor.
     *
     * @param int|null $val The value to be stored in the node.
     * @param ListNode|null $next The next node in the linked list.
     */
    public function __construct(null|int $val = 0, ListNode $next = null)
    {
        $this->val = $val;
        $this->next = $next;
    }
}

/**
 * Class Solution
 *
 * This class provides a solution for doubling the value of each node in a linked list.
 */
class Solution
{
    /**
     * Doubles the value of each node in a linked list.
     *
     * The function iterates through the linked list, doubling the value of each node.
     * If the value of a node is greater than or equal to 5, a new node with a value of 0 is added before it.
     *
     * @param ListNode $head The head of the linked list.
     * @return ListNode The head of the modified linked list.
     */
    public function doubleIt(ListNode $head): ListNode
    {
        if ($head->val >= 5) {
            $head = new ListNode(0, $head);
        }
        $current = $head;
        while ($current->next) {
            $current->val = ($current->val * 2) % 10;
            if ($current->next->val >= 5) {
                $current->val++;
            }
            $current = $current->next;
        }
        $current->val = ($current->val * 2) % 10; // Handle the last node
        return $head;
    }
}

File: dart/leet_code/2816.dart
/// Copyright ©2012-2024 JerodG <https://github.com/jerodg/>
///
/// This program is free software: you can redistribute it and/or modify it under the terms of the
/// Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
/// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
/// for more details.
///
/// The above copyright notice and this permission notice shall be included in all copies or
/// substantial portions of the Software. You should have received a copy of the SSPL along with this
/// program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.
library;

/// A file that defines a linked list node and a solution to double the value of each node in the list.

/// A class that represents a node in a singly linked list.
/// Each node has an integer value and a reference to the next node in the list.
class ListNode {
  /// The integer value of the node.
  int val;

  /// The next node in the list.
  ListNode? next;

  /// Creates a new node with the given value and next node.
  /// If no value is provided, the default is 0.
  /// If no next node is provided, the default is null.
  ListNode([this.val = 0, this.next]);
}

/// A class that provides a solution to double the value of each node in a linked list.
class Solution {
  /// Doubles the value of each node in the given linked list.
  /// If the doubled value is greater than 9, carries the value to the next node.
  /// If the value of the first node is carried, adds a new node with value 1 at the beginning of the list.
  ///
  /// @param head The first node of the linked list.
  /// @return The first node of the modified linked list.
  ListNode? doubleIt(ListNode? head) {
    final firstDigit = dfs(head);

    if (firstDigit == 0) {
      return head;
    }

    return ListNode(1, head);
  }
}

/// Doubles the value of the given node and all following nodes in the list.
/// If the doubled value is greater than 9, carries the value to the next node.
///
/// @param cur The current node to double.
/// @return 1 if the doubled value of the current node is greater than 9, otherwise 0.
int dfs(ListNode? cur) {
  if (cur == null) {
    return 0;
  }

  final newVal = cur.val * 2 + dfs(cur.next);
  cur.val = newVal % 10;

  return newVal > 9 ? 1 : 0;
}
